Start dynamic heap count work

PeterSolMS opened the pull request:
This is work to enable changing the heap count for server GC dynamically:
- move regions between heaps
- make handle table scanning accommodate varying numbers of GC threads
- changes to mark list and mark list piece sizing
- rethreading free lists for gen 2 (other generations to follow)
- update thread numbers in joins to make joins work correctly
- park idle GC threads on a new event `gc_idle_thread_event`
- changes to allow `try_allocate_more_space` to switch to a different heap
- move finalization data between heaps

Note that with the current code we are always running in a stress mode where we vary the heap count randomly every 100 GCs.

PeterSolMS committed:
Implementation of redistribute_regions along with some hacks to enable a bit of testing.

PeterSolMS committed:
Implementation of redistribute_regions along with some hacks to enable a bit of testing.

PeterSolMS committed:
More changes to support dynamic heap count:
- park extra threads on gc_idle_thread_event
- update thread count in join
- null free lists and have background GC rebuild them
- have redistribute_regions call fix_allocation_contexts
- distribute free regions as well
- fix up ephemeral_heap_segment and generation_allocation_segment

PeterSolMS committed:
More changes to support dynamic heap count:
- park extra threads on gc_idle_thread_event
- update thread count in join
- null free lists and have background GC rebuild them
- have redistribute_regions call fix_allocation_contexts
- distribute free regions as well
- fix up ephemeral_heap_segment and generation_allocation_segment

PeterSolMS committed:
Incorporate handle table changes, fix mark_list_piece, mark_list sizing.

Add hack to dynamically enable heap verify when we change the heap count.

PeterSolMS committed:
Incorporate handle table changes, fix mark_list_piece, mark_list sizing.

Add hack to dynamically enable heap verify when we change the heap count.

PeterSolMS committed:
rethreading freelists

PeterSolMS committed:
rethreading freelists

PeterSolMS committed:
Merge branch 'main' into redistribute_regions

PeterSolMS committed:
fixed #ifdef for segments version

PeterSolMS committed:
fixed #ifdef for segments version

PeterSolMS committed:
Enhancements:
- move finalization data between heaps
- update free list space per heap when rethreading free lists
- update allocation contexts so they don't reference decommissioned heaps
- allow redistribute_regions to fail
- allow enter_spin_lock to fail when called from try_allocate_more_space
- don't decommision heaps with a taken more space lock
- poison dynamic data and generation table for decommissioned heaps

PeterSolMS committed:
Enhancements:
- move finalization data between heaps
- update free list space per heap when rethreading free lists
- update allocation contexts so they don't reference decommissioned heaps
- allow redistribute_regions to fail
- allow enter_spin_lock to fail when called from try_allocate_more_space
- don't decommision heaps with a taken more space lock
- poison dynamic data and generation table for decommissioned heaps

PeterSolMS committed:
Rethread free lists for the other generations.

PeterSolMS committed:
Rethread free lists for the other generations.

PeterSolMS committed:
- Fix free list rethreading issue
- Add instrumentation
- Be more careful regarding signed vs. unsigned types to make GCC happy.

PeterSolMS committed:
- Fix free list rethreading issue
- Add instrumentation
- Be more careful regarding signed vs. unsigned types to make GCC happy.

PeterSolMS committed:
Fixes for GCC compiler.

PeterSolMS committed:
Fixes for GCC compiler.

PeterSolMS committed:
Rename restribute_regions to change_heap_count, refactor code, add checking for containing heap to verify_free_lists, adding checking decommissioned heaps to verify_heap.

PeterSolMS committed:
Rename restribute_regions to change_heap_count, refactor code, add checking for containing heap to verify_free_lists, adding checking decommissioned heaps to verify_heap.

PeterSolMS committed:
Merge branch 'main' into redistribute_regions

PeterSolMS committed:
Fix compile error with segments.

PeterSolMS committed:
Fix compile error with segments.

Commit:
misc change to get repros

Commit:
misc change to get repros

Commit:
got rid of code no longer needed

Commit:
got rid of code no longer needed

Commit:
different way of handling decommissioned heaps

Commit:
different way of handling decommissioned heaps

Commit:
do not set some fields to 0 when recommissioing a heap as we concurrently cannot handle that

Commit:
do not set some fields to 0 when recommissioing a heap as we concurrently cannot handle that

Commit:
enabled msl and loh alloc state recording

Commit:
enabled msl and loh alloc state recording

PeterSolMS committed:
Compile coreclr.dll with regions, enable memory mapped stresslog

PeterSolMS committed:
Compile coreclr.dll with regions, enable memory mapped stresslog

PeterSolMS committed:
Merge with main.

PeterSolMS committed:
Fix compilation error after merge.

PeterSolMS committed:
Fix compilation error after merge.

PeterSolMS committed:
Merge branch 'redistribute_regions' of https://github.com/petersolMS/runtime-1 into redistribute_regions

PeterSolMS committed:
Merge with main, cleanup, remove demotion related changes.

PeterSolMS committed:
Merge branch 'main' into redistribute_regions

PeterSolMS committed:
Small change to make spin lock logic clearer.

Maoni0 commented:
>     gc_join_max = 40

this should be deleted since you are not using it.

---
Refers to: src/coreclr/gc/gc.cpp:704 in fa4cf11. [](commit_id = fa4cf11a6c39eb6119db8a7033e3b627330003cc, deletion_comment = False)

Maoni0 commented on src/coreclr/debug/inc/amd64/primitives.h:
```diff
@@ -70,7 +70,9 @@ constexpr CorDebugRegister g_JITToCorDbgReg[] =
     REGISTER_AMD64_R13,
     REGISTER_AMD64_R14,
-    REGISTER_AMD64_R15
+    REGISTER_AMD64_R15,
+    REGISTER_AMD64_RSP,
+    REGISTER_AMD64_RSP,
```
is this related to this PR?

PeterSolMS replied:
No, this is just a temporary hack to work around a debugging issue I encountered - I'll take it out.

Maoni0 commented on src/coreclr/gc/gcinterface.h:
```diff
@@ -1032,6 +1032,7 @@ struct ScanContext
 {
     Thread* thread_under_crawl;
     int thread_number;
+    int thread_count;
```
this would break the backward compatibility with the VM side (ie, you can't run clrgc.dll with a previous version of coreclr.dll anymore). can you change it in a way that does not introduce this breaking change?

Maoni0 commented on src/coreclr/gc/gcpriv.h:
```diff
@@ -142,6 +142,14 @@ inline void FATAL_GC_ERROR()
 #define USE_REGIONS
 #endif //HOST_64BIT && BUILD_AS_STANDALONE
 
+#define SPINLOCK_HISTORY
+#define RECORD_LOH_STATE
```
```suggestion
// #define SPINLOCK_HISTORY
// #define RECORD_LOH_STATE
```

Maoni0 replied:
you wouldnt want to define these unconditionally - at least not in release builds

PeterSolMS replied:
Will do.

Maoni0 commented on src/coreclr/gc/gcpriv.h:
```diff
@@ -347,7 +355,8 @@ const int policy_expand  = 2;
 
 void GCLog (const char *fmt, ... );
-#define dprintf(l,x) {if ((l == 1) || (l == GTC_LOG)) {GCLog x;}}
+//#define dprintf(l,x) {if ((l == 1) || (l == GTC_LOG)) {GCLog x;}}
+#define dprintf(l,x) {if ((l == 1) || (l == 5555)) {GCLog x;}}
```
```suggestion
#define dprintf(l,x) {if ((l == 1) || (l == GTC_LOG)) {GCLog x;}}
```

PeterSolMS replied:
Sure...

Maoni0 commented on src/coreclr/gc/gcpriv.h:
```diff
@@ -571,9 +580,17 @@ enum allocation_state
     a_state_trigger_ephemeral_gc,
     a_state_trigger_2nd_ephemeral_gc,
     a_state_check_retry_seg,
+    // a temp state I added to distinguish 
+    a_state_wait_in_tams,
```
did you want to get rid of this?

PeterSolMS replied:
Yes, this seems low value at this point.

Maoni0 commented on src/coreclr/gc/gcpriv.h:
```diff
@@ -3927,13 +4013,19 @@ class gc_heap
 #endif //MH_SC_MARK
 
+#ifdef USE_REGIONS
+    PER_HEAP_ISOLATED_FIELD_SINGLE_GC size_t total_num_fl_items_stage1;
+    PER_HEAP_ISOLATED_FIELD_SINGLE_GC size_t total_num_fl_items_moved_stage1;
```
I wrote these for verification purposes. you could get rid of them.

Maoni0 commented on src/coreclr/gc/gcpriv.h:
```diff
@@ -4391,6 +4485,10 @@ class gc_heap
 #endif //HEAP_ANALYZE
 
+#ifdef MULTIPLE_HEAPS
+    PER_HEAP_FIELD bool uoh_msl_before_gc_p;
```
this should be under the `#ifdef STRESS_DYNAMIC_HEAP_COUNT` check shoiuldn't it?

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -1376,6 +1385,85 @@ static void safe_switch_to_thread()
+
+// If our heap got decommissioned, we need to try an existing heap.
+//inline
+#pragma optimize("", off)
```
```suggestion
```

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -1376,6 +1385,85 @@ static void safe_switch_to_thread()
+    return false;
+#endif //MULTIPLE_HEAPS
+}
+#pragma optimize("", on)
```
```suggestion
```

Maoni0 replied:
I turned the optimizatio off for debugging purpose. should not be checked in.

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -1385,11 +1473,12 @@ static void enter_spin_lock_noinstru (RAW_KEYWORD(volatile) int32_t* lock)
 retry:
 
-    if (Interlocked::CompareExchange(lock, 0, -1) >= 0)
+    if (Interlocked::CompareExchange(lock, lock_taken, lock_free) != lock_free)
```
this method doesn't actually need to change, correct? it will work as is. but I do think it's better to use the actualy enum instead of an integer. perhaps add a comment to say that this will never be used by the more space lock?

PeterSolMS replied:
Right, this doesn't need a change, I just did it for clarity. Will add a comment as suggested.

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -6885,11 +7032,26 @@ void gc_heap::gc_thread_function ()
+                    leave_spin_lock(&hp->more_space_lock_soh);
+                }
+
+                if ((lowest_heap_with_msl_uoh == -1) && (hp->uoh_msl_before_gc_p))
+                {
+                    lowest_heap_with_msl_uoh = i;
+                }
+
+                if (hp->uoh_msl_before_gc_p)
+                {
+                    dprintf (5555, ("h%d uoh msl was taken before GC", i));
+                    hp->uoh_msl_before_gc_p = false;
+                }
```
shouldn't this code be under `#ifdef STRESS_DYNAMIC_HEAP_COUNT`?

PeterSolMS replied:
Yes, it should be.

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -1581,19 +1676,24 @@ static void enter_spin_lock (GCSpinLock* spin_lock)
 
 #endif //_DEBUG
 
+inline static BOOL spin_lock_taken_p (GCSpinLock* pSpinLock)
```
do we need to 2 `spin_lock_taken_p` functions?

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -2215,6 +2319,10 @@ size_t*     gc_heap::g_bpromoted;
 
 BOOL        gc_heap::gradual_decommit_in_progress_p = FALSE;
 size_t      gc_heap::max_decommit_step_size = 0;
+#ifdef USE_REGIONS
+size_t      gc_heap::total_num_fl_items_moved_stage1 = 0;
+size_t      gc_heap::total_num_fl_items_stage1 = 0;
+#endif //USE_REGIONS
```
```suggestion
```

PeterSolMS replied:
Sure.

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -2225,6 +2333,8 @@ size_t      gc_heap::g_promoted;
 size_t      gc_heap::g_bpromoted;
 #endif //BACKGROUND_GC
 
+const int n_heaps = 1;
```
can you please add an explanation why this is called the same name as the one in `gc_heap`?

PeterSolMS replied:
I added a comment to that effect as well.

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -6217,7 +6339,12 @@ class heap_select
         {
             uint32_t proc_no = GCToOSInterface::GetCurrentProcessorNumber();
-            return proc_no_to_heap_no[proc_no];
+            int best_heap = proc_no_to_heap_no[proc_no];
```
I wouldn't call this `best_heap` which makes it sound like this is just a perf thing. but it wouldn't work if we didn't do `%`. probably call it something like `adjusted_heap`.

PeterSolMS replied:
ok.

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -23562,7 +24282,21 @@ void gc_heap::equalize_promoted_bytes()
+            if (heap_segment_rw (generation_start_segment (gen)) == nullptr)
+            {
+                size_t size = gen_idx > max_generation ? global_region_allocator.get_large_region_alignment() : 0;
+                heap_segment* start_region = hp->get_free_region (gen_idx, size);
```
this can fail - at the minimum we should add a TODO here.

PeterSolMS replied:
I have added a comment.

PeterSolMS replied:
I rearranged the code so we check before whether we have or can get enough regions for each generation **before** we commit to increasing the heap count.

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -23578,8 +24312,677 @@ void gc_heap::equalize_promoted_bytes()
+#endif //RECORD_LOH_STATE
+}
+
+bool gc_heap::change_heap_count (int new_n_heaps)
```
it would be great if you could add some detailed comments in this method like for each thing we are adjusting, what we are actually doing, like what we do in `thread_final_regions` or `equalize_promoted_bytes`, eg

```
// step 1: adjust finalize queue. if the heap count is larger, [describe how we split the old heaps' finalize queue into more queues]; if it's smaller, [describe how we decide to merge which heap into which heap]
// step 2: for all existing regions, prepare the regions by pretending all their allocated space survives so equalize_promoted_size will distribute regions evenly across new heaps
// ...
```
so we don't just say "adjust data for new heaps" but rather describe how we are adjusting - for example, if we go from 10 to 6, are we adjusting by taking the last 4 heaps and move all their regions to the 1st 4 heaps? or are we distributing evenly to all 6 heaps?

PeterSolMS replied:
Will do...

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -23578,8 +24312,677 @@ void gc_heap::equalize_promoted_bytes()
+            to_heap_number = (to_heap_number + 1) % new_n_heaps;
+        }
+    }
+
```
from this point on, is it true we cannot fail? (it's actually not true since we could fail in equalize_promoted_bytes. is it ok do all the steps that can fail before doing steps that can't?

PeterSolMS replied:
I think we absolutely need to avoid the case where we get stuck because we failed in the middle of adjusting the heap count, but cannot go back. One way to avoid that is to do all the steps that can fail first, and make sure we can recover. I think this is feasible, but of course you are correct in pointing out it's curently not the case.

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -23578,8 +24312,677 @@ void gc_heap::equalize_promoted_bytes()
+    // even out the regions over the current number of heaps
+    equalize_promoted_bytes (max_generation);
+
+    total_num_fl_items_moved_stage1 = ~0;
+    total_num_fl_items_stage1 = ~0;
+
```
```suggestion
```

PeterSolMS replied:
Sure, I'll remove the declarations and uses.

Maoni0 commented on src/coreclr/gc/gc.cpp:
```diff
@@ -23578,8 +24312,677 @@ void gc_heap::equalize_promoted_bytes()
+    // make sure no allocation contexts point to idle heaps
+    fix_allocation_contexts_heaps();
+
+    GCConfig::s_HeapVerifyLevel = 1;
```
you could just call verify_heap here instead of turning it on for all GCs later as we discussed (or verify this one and a couple of GCs later

PeterSolMS replied:
Hmm, this code is on a single thread, so I'd have to iterate over all heaps. Needs more thought.

Maoni0 replied:
ahh right, one idea is to set a global var to indicate that we did change the HC (heap count) like -

`PER_HEAP_ISOLATED_FIELD_SINGLE_GC bool last_suspension_hc_changed_p; `

and next time we do a GC, we will call `verify_heap` on entry if this is true and that GC will reset it. or you could keep a var to record the GC index when we suspended to change the HC, and call `verify_heap` in the next few GCs if this is not -1.

PeterSolMS replied:
Actually, in the dynamic heap count PR building on this one, I execute the `change_heap_count` method on all GC threads, and so it would be easy to just call `verify_heap` at the end.

PeterSolMS committed:
Address code review feedback.

PeterSolMS committed:
Address code review feedback:
- improve comments
- make sure we check first whether we have or can allocate enough regions to increase heap count before committing to the increase

PeterSolMS committed:
Undo white space change

PeterSolMS committed:
Improve comments, rearrange code for clarity - put the steps first that can cause us to fail changing heap count.

PeterSolMS committed:
Port 3 bug fixes from dynamic_heap_count branch.

PeterSolMS committed:
Remove extra comma inserted accidentally.

PeterSolMS committed:
Enhance comments in change_heap_count.

PeterSolMS committed:
Merge with main

PeterSolMS committed:
Fix GCC warning messages.

PeterSolMS committed:
More GCC warning work.

PeterSolMS commented:
Closing this pull request as it has been superseded by the dynamic heap count one.
